File: rotate-list/rotate-list.py
class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head or not head.next or k==0:
            return head
        slow=head
        fast=head 
        count=0
        flag=0
        while fast and fast.next:
            count+=1
            slow=slow.next
            if not fast.next.next:
                fast=fast.next
                flag=1
                break 
            else:
                fast=fast.next.next 
        if flag==1:
            l = 2*count 
        else:
            l=2*count+1
        p=head 
        step=(l-k%l)%l
        if step==0:
            return head
        while step>0:
            step-=1
            q=p
            p=p.next
        q.next=None    
        fast.next=head
        return p

File: 61-rotate-list/rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */


class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || !head->next || k == 0) return head;

        //find length and last node
        int n = 1;
        ListNode* tail = head;
        while (tail->next) {
            tail = tail->next;
            n++;
        }

        // reduce k
        k = k % n;
        if (k == 0) return head;

        // make circular
        tail->next = head;

        // find new tail (n - k - 1 steps)
        int steps = n - k;
        ListNode* newTail = head;
        while (--steps) {
            newTail = newTail->next;
        }

        // break circle
        ListNode* newHead = newTail->next;
        newTail->next = nullptr;

        return newHead;
    }
};